# fusionfusion 

## Introduction

fusionfusion is a software for detecting gene fusion 
using the putative chimeric transcript generated by several well-known transcriptome alignment tools (STAR, MapSplice 2).
Although many of predicted chimeric transcripts are are "false positives",
by performing effective filtering, sensitive and accurate gene fusion detection is possible.
After the alignment step, the software can generate the final gene fusion candidates within about one hour,
and integrating our software into the pipeline will come very easily to you!

## Dependency

### Python
Python (>= 2.7), `pysam (>= 0.8.1)`, `pyyaml` packages

### Software
blat

## Install

```
git clone https://github.com/Genomon-Project/fusionfusion.git
cd fusionfusion 
python setup.py build
python setup.py install
```
## Preparation

First, you need to perform transcriptome sequencing alignemnt by STAR or MapSplice 2 
(we will soon support Tophat2).

For STAR, our software uses the chimeric sam file
```
Chimeric.out.sam
```

For MapSplice 2, our software uses the read alignment file
```
alignments.sam (bam)
```
You do not need to care about the sorting status.

Then, a configuration files (in yaml format) should be prepared.
```
param.yaml
```

See sample files for description of each parameters.

## Commands


```
fusionfusion --star star_chimera.bam --ms2 ms2.bam --out output_dir --param param.yaml
```

## Results

1. chromosome for the 1st breakpoint
1. coordinate for the 1st breakpoint
1. direction of the 1st breakpoint
1. chromosome for the 2nd breakpoint
1. coordinate for the 2nd breakpoint
1. direction of the 2nd breakpoint
1. inserted nucleotides within the breakpoints
1. type of the structural variation
1. gene overlapping the 1st breakpoint
1. gene overlapping the 2nd breakpoint
1. exon overlapping the 1st breakpoint
1. exon overlapping the 2nd breakpoint
1. #read_pairs not supporting the variant (reference read pairs) for the tumor sample
1. #read_pairs supporting the variant (variant read paris) for the tumor sample
1. #read_pairs not supporting the variant for the matched control sample
1. #read_pairs supporting the variant for the matched control sample
1. frequency of variant read pairs for the tumor sample 
1. frequency of variant read pairs for the matched control sample 
1. p-value for the Fisher's exact text (on contingency table of (tumor v.s. matched control) and (reference v.s. variant read pairs)


